#### UNDERSTANDING PYTHON PROGRAMMING $\;\;\;\;\;\;\;\;\;\;\;$   Created By: Siddhant Patel  $\;\;\;\;\;\;\;\;\;\;\;$   https://github.com/sid-patel 

# Chapter 1 : Key Concepts in Python

### 1. Variable Assignment in Python
### 2. Memory Model in Python
### 3. Dynamic Typing in python
### 4. Immutable vs Mutable Objects in Python
### 5. Namespaces and Variable Scope
### 6. Unerstanding "self" in Python
### 7. Understandin __ init __ method in python
### 8. Understanding design Philosophy in Python - "Everything in Python is an Object".

 ## 1. Variable Assignment in Python

*  Variable in python is essentially a reference/name/label assigned or given to piece of data.

*  '=' is assignment operator. it means if age = 15 then 15 (value or data) is assigned (or stored) to variable called 'age'.

*  When we declare a variable in python, first python creates an object in memory to represent value (e.g. 15), then variable "age" becomes reference to that object. 

##### Do Variables Hold Memory?
##### Technically, variables themselves do not hold memory in Python. Instead:

* **Values or objects** are stored in memory.
* Variables are references (or pointers) to those memory locations.

##### If multiple variable reference the same object, then they share same memory until one of them is re-assigned. Shown in below example.

In [15]:
age = 15
print(age)
print(id(age)) # id is memory location of variable "age". so "age" is a variable(reference) that is pointing id(location) where object (or value) 15 is stored.

15
140721211413736


In [1]:
a = 10
b = 10
print(id(a))
print(id(b))

140721571206216
140721571206216


* Note : so as we can see here both the Variables a and b reference the same object 10 so they share the same memory location

##### Initializing multiple variables

In [6]:
Name,Id,Age = "Siddhant",1001,30
print("Name:",Name)
print("Id:",Id)
print("Age:",Age)

Name: Siddhant
Id: 1001
Age: 30


 ## 2. Memory Model in Python

  #### Python uses a reference counting mechanism for memory management. This is part of Python's Garbage Collection (GC) system. Here's how it works:
 * Every object in Python has a reference count (how many variables point to it).
 * When an object’s reference count drops to zero (i.e., no variable points to it), Python automatically frees that memory.

In [1]:
x = [1, 2, 3]
y = x       # `y` references the same list object
print(id(x))  # Same memory address for `x` and `y`
print(id(y))

1740289889024
1740289889024


In [3]:
y = None     # Now `y` no longer points to the list
print(x)     # `x` still references the list
print(y)
print(id(x)) 
print(id(y))

[1, 2, 3]
None
1740289889024
140721210026736


 ## 3. Dynamic Typing in Python
 #### Python is dynamically typed, meaning:

 * A variable can reference different types of objects at different times.
 * You don’t need to declare the type of a variable explicitly.

In [7]:
x = 25
print(x)
print(type(x))
print(id(x)) 

x = ('hello')
print(type(x))
print(id(x)) 
print(x)

25
<class 'int'>
140721211414056
<class 'str'>
1740268689456
hello


#### In Above code :
* first we assigned value 25 (integer) to variable x so for that data type of x is <class 'int'>
* then by reassigning x we changed value of x = 'hello' (string) so for that data type of x is <class 'str'>
#### Reassignment :
* When x is reassigned from 25 to 'hello', Python creates a new object for 'hello' in memory (at 1740268689456), and the reference to 25 is discarded (if nothing else references 25). id(x) changes because x points to a new object.

#### Dynamic Typed :
* Therefore variables are dynamically typed, they dont have data types but their data types depend upon data assigned / or stored in them. For example in c or c++ you assign data type while assigning variable ex. int x = 10 so here they are static. their data type wont change.
* Python is dynamically typed, meaning a variable can change its type during execution. Initially, x was an integer (int), but after reassignment, it became a string (str).


#### What Happens to Memory Location of 25?
* When x is reassigned to a new object ('hello'), the previous object (25) is unaffected in memory. However, its behavior depends on how many references are pointing to it.
##### 1. Reference count : 
##### In Python, objects are managed using a mechanism called reference counting:

* If no other variables or references point to 25, its reference count becomes 0.
* At this point, Python’s garbage collector marks the object (25) for cleanup, and its memory is reclaimed eventually.
* If another variable still points to 25, the object remains in memory.

##### Demonstration of Reference Count:
* You can manually inspect the reference count of an object using the sys.getrefcount() function:

In [11]:
import sys

print(sys.getrefcount(25)) # checking reference count for object or value 25.

x = 25
print(sys.getrefcount(25))  # Reference count for 25 increased by one as we assign one more reference x to the object 25. ( reference is usually higher due to internal references)

x = 'hello'                 # x no longer points to 25
print(sys.getrefcount(25))  # Reference count decreases by one as x is no longer points 25 because x points "hello" now. so reference counnt is decreased if no other references exist


1000000047
1000000048
1000000047


##### 2. Garbage Collection:

* When the reference count for an object drops to zero, **Python’s garbage collector** eventually cleans it up. This is not immediate but happens during the next garbage collection cycle.
* Python uses an internal garbage collection system to free up unused memory.


## 4. Immutable vs. Mutable Objects
* **Immutable objects (e.g., int, float, tuple, str)** cannot be changed after they are created. When you modify an immutable object, Python creates a new object in memory.
* **Mutable objects (e.g., list, dict, set)** can be changed without creating a new object.

#### Example : Immutable 

In [4]:
x = 10       # once integer 10 is assigned to x, it cant be changed
x += 1       # A new object is created for 11
print(x)

11


#### Example : Mutable

In [3]:
x = [1, 2, 3]
print(x)
print(id(x))
x.append(4)  # same list object can be modified and here append Modifies the list in-place
print(x)
print(id(x))

[1, 2, 3]
2259248504128
[1, 2, 3, 4]
2259248504128


## 5. Namespaces and Variable Scope
#### A namespace is a mapping of variable names to objects. Python has different types of namespaces:

* **Global Namespace:** Variables defined at the module level.
* **Local Namespace:** Variables defined inside functions.
* **Built-in Namespace:** Includes Python's built-in functions and objects.

In [35]:
x = 5  # Global variable

def Func():                  # defining function 
    x = 10  # Local variable
    print("Inside Func value of x is :", x)
    print(id(x))

s1 = Func()                       # calling function # we created s1 instance for it.
print("Outside Func value of x is :", x)
print(id(x))

Inside Func value of x is : 10
140728534602824
Outside Func value of x is : 5
140728534602664


* **Above Program Execution Flow**:
1. We assigned value 5 to variable x which is our Global Variable.
2. We have definded function called Func using inbuit keyword def that has variable x with value 10 and two print statements one by one.
3. As we are calling our Func() so Func() will be called and executed the output of execution is first two lines in our output.
4. Then both the print statements will be executed that are outside of func().

In [36]:
print(s1)

None


#### Key Observations:
**1. Global x :**
* The variable x = 5 is defined at the **global scope**, so it is accessible anywhere in the code outside the function Func().

**2. Local x :**

* The variable x = 10 is defined inside the Func() function, so it is a local variable and exists only during the function’s execution.
* **The local variable x inside the Func() function is not accessible outside the function.** This is because local variables have a **function scope** — they only exist and are accessible during the execution of the function in which they are defined. Once the function finishes execution, the local variable x is destroyed and cannot be accessed from outside.

**3. Function Return:**

* The function Func() does not return the local x, so **s1** is assigned the return value of **None** (default return value for a function with no return statement).


#### How to access Local x ?
##### Option 1: Return the Local Variable
* Modify the function to return the local x:

In [39]:
x = 5  # Global variable

def Func2():                  # defining function 
    x = 10  # Local variable
    print("Inside Func2 value of x is :", x)
    print(id(x))             # it will return MEMORY ID OF LOCAL x as an output
    return x                 # Return the Local variable

s2 = Func2()                  # s2 now stores the value of local x
print("Outside Func2 value of x is :", x)

Inside Func2 value of x is : 10
140728534602824
Outside Func2 value of x is : 5


In [40]:
print("Value of Local x stored in s2 is :",s2)

Value of Local x stored in s2 is : 10


##### Option 2: Store the Local Variable in a Class Attribute
* We can use a class to encapsulate the local variable as an attribute:

In [56]:

x = 5  # Global variable

class MyClass:
    def Func3(self):
        self.x = 10  # Store local x in an instance attribute
        print("Inside Func3 value of x is :", self.x)
        print("ID of self.x is:",id(self.x))

s3 = MyClass()
s3.Func3()

print("Access Local x via s3.x:", s3.x)
print("ID of Local x via s3.x:", id(s3.x))
print("Outside of Myclass value of global x is:",x)

Inside Func3 value of x is : 10
ID of self.x is: 140728534602824
Access Local x via s3.x: 10
ID of Local x via s3.x: 140728534602824
Outside of Myclass value of global x is: 5


In [57]:
s4 = MyClass()
s4.Func3()
print("ID of Local x via s4.x:", id(s4.x))

Inside Func3 value of x is : 10
ID of self.x is: 140728534602824
ID of Local x via s4.x: 140728534602824


##### Note:
* In Output **id(self.x)** ---> **140728534602824**  and **id(s3.x)** is also -----> **140728534602824**. It means they are same. **self.x = self.s3**
* It means **self = s3** for our s3 object (instance) when we assign or create **s3 = MyClass()**
  so now inside Func3 scenario is **s3.x = 10** so s3.x variable has value 10 so when we print(s3.x) output is 10
* Also when we create instance **s4 = Myclass()** and call function using **s4.Func3()** in that case our **self.x = s4.x = 10** because we can see
  the memory ID of both are same.
* Another interesting thing is **s3.x** and **s4.x** both are **different instances of MyClass has same Memory ID** ---> **140728534602824**. It is because the value stored is same which is 10. for example, when we assign a=10 , b = 10 then both a and b references holds same object 10 so both will point the same memory location where 10 is stored. its called **multiple variables assigned to same object**.(explained above in variable assignment) 

#### Lets see What happens when we change value of Local variable self.x for different instances by giving user input :

In [62]:
class MyClass2:
    def Func4(self):
        self.x = input("Provide value for Local x :")  # taking value of local x from user for every new instance of class
        self.y = input("Provide value for Local y :")
        print("Inside Func4 value of self.x is :", self.x)
        print("Inside Func4 value of self.y is :", self.y)
        print("ID of self.x is:",id(self.x))
        print("ID of self.y is:",id(self.y))
        
s5 = MyClass2()
s5.Func4()

print("Access Local self.x via s5.x:", s5.x)
print("Access Local self.y via s5.y:", s5.y)
print("ID of s5.x:", id(s5.x))
print("ID of s5.y:", id(s5.y))

Provide value for Local x : 15
Provide value for Local y : 20


Inside Func4 value of self.x is : 15
Inside Func4 value of self.y is : 20
ID of self.x is: 2259265753456
ID of self.y is: 2259265754544
Access Local self.x via s5.x: 15
Access Local self.y via s5.y: 20
ID of s5.x: 2259265753456
ID of s5.y: 2259265754544


In [64]:
s6 = MyClass2()
s6.Func4()
print("Access Local self.x via s6.x:", s6.x)
print("Access Local self.y via s6.y:", s6.y)
print("ID of s6.x:", id(s6.x))
print("ID of s6.y:", id(s6.y))

Provide value for Local x : 30
Provide value for Local y : 50


Inside Func4 value of self.x is : 30
Inside Func4 value of self.y is : 50
ID of self.x is: 2259265861104
ID of self.y is: 2259265859632
Access Local self.x via s6.x: 30
Access Local self.y via s6.y: 50
ID of s6.x: 2259265861104
ID of s6.y: 2259265859632


#### Note :
* For our s5 instance **self** will become **s5** so during **s5 = MyClass(), s5.func4()** we are **replacing self with s5** thats 
  why **id of self.x = id of s5.x** and **id of self.y = id of s5.y**
* Same way, for our s6 instance **self** will become **s6** so during **s6 = MyClass(), s6.func4()** we are **replacing self with s6** thats 
  why **id of self.x = id of s6.x** and **id of self.y = id of s6.y**
##### Here's importance of self :
* Using self we can create as many attributes as we want like self.x,self.y,self.g and so on. and with single instance for example s5 we will be able   to access all of them

#### Understanding "self" inbuilt keyword in Python with Real World Scenario :
* **Problem Statement :** Assume you're a teacher having 3 students in your class. 
* You are teaching them Math,Science and Physics and taking their monthly test for each subject of 100 Marks(each test).
* so after monthly exams, you have a sheet with 5 Columns. 
* In 1st column there are their names (David,Keval,Sid).
* In 2nd column "Obtained marks in Math out of 100"
* In 3nd column "Obtained marks in Science out of 100"
* In 4th column "Obtained marks in Physics out of 100"
* In 5th column is of "Percentage" that what you want to calculate using python.
* **Goal** is to calculate Percentage(%) for each student so whoever got highest percentage will be rewarded.
* So we will create a python program that will take an input Name,Math Score,Science Score,Physics Score and
  will give us percentage as an output for that student. so we will do the same for 3 times as we have five students.

#### 1st Version : Basic Program using Class and method (percentage)
##### Scope of improvement :
* 1.In this below version there is no error handling. it means while entering student data if user makes mistakes and enters string or -ve number etc
  then this kinds of erros aren't handled by this program.
* 2.Also note that for each student we created instance like student1,student2,student3 so we are executing whole program 3 time one by one using each instance to calculate percentage. insted of that we can design it in a way that it takes all 3 student input once and execute program once and give an output for all 3 at once.
* These above 2 things we will do in our **2nd Version** and **3rd Version**of Program.

In [ ]:
class Basic_Student:
    def percentage(self):
        # Displaying the type of self
        print(f"The data type of 'self' is: {type(self)}")

        # Taking user inputs
        self.name = input("Enter the student's name: ")
        self.maths = float(input("Enter marks for Maths (out of 100): "))
        self.science = float(input("Enter marks for Science (out of 100): "))
        self.english = float(input("Enter marks for English (out of 100): "))

        # Calculating percentage
        total_marks = self.maths + self.science + self.english
        percentage = (total_marks / 300) * 100

        # Displaying the result
        print(f"\n{self.name}'s Percentage: {percentage:.2f}%")

# Creating an instance of the Student class
student1 = Basic_Student()
student2 = Basic_Student()
student3 = Basic_Student()

In [89]:
# Calling the percentage method
student1.percentage()
print(id(student1.name))   # when we run this method for student1 instance then our self becomes student1. self=student1
print(id(student1.maths))  # it means inside our __init__function variables will be student1.name,student1.maths,student1.science,student1.physics

Enter the student's name:  David
Enter score for Maths (out of 100):  78
Enter score for Science (out of 100):  87
Enter score for Physics (out of 100):  67



David's Percentage: 77.33%
2259253653488
2259261765008


In [92]:
student2.percentage() # calling percentage method for 2nd student (student2 instance)
print(id(student2.name))   # when we run this method for student2 instance then our self becomes student2. self=student2
print(id(student2.maths))

Enter the student's name:  Keval
Enter score for Maths (out of 100):  56
Enter score for Science (out of 100):  89
Enter score for Physics (out of 100):  96



Keval's Percentage: 80.33%
2259265765296
2259261760784


In [91]:
student3.percentage() # calling percentage method for 3rd student (student3 instance)
print(id(student3.name))   # when we run this method for student3 instance then our self becomes student3. self=student3
print(id(student3.maths))

Enter the student's name:  Sid
Enter score for Maths (out of 100):  78
Enter score for Science (out of 100):  67
Enter score for Physics (out of 100):  88



Sid's Percentage: 77.67%
2259264750960
2259261763536


## 6. Understanding "Self" in Python :

* In Python, **self** is a conventionally used parameter name in instance methods, including the **__ init __** method, to refer to the instance of the class that is calling the method. It acts as a reference to the object itself and allows access to its attributes and methods.

##### Definition of self :
* **self is a reference to the current instance of the class and is used to access variables and methods that belong to the class instance.**

#### Key Points about self:
**1. Instance Access:** It allows access to the attributes and methods of the current instance.

**2. Mandatory First Parameter:** For instance methods, self must be the first parameter (though you can name it something else, it is highly discouraged to break this convention).

**3. Scope:** self is only relevant inside instance methods, where it refers to the specific object on which the method is being invoked.

##### Why is self Explicit?
* Python uses self explicitly in method definitions to make it clear which attributes and methods belong to the instance. This explicitness avoids ambiguity and enhances code readability.
* In summary, **self** serves as a bridge between the methods and attributes of an instance and the actual instance object calling the method. It's crucial for object-oriented programming in Python!

#### Explaination of Syntax of Print Statement used in above program:
* #### print(f"\n{self.name}'s Percentage: {percentage:.2f}%")
* ##### Explanation:
1. **f" at the Start:** The f" indicates that this is an f-string, allowing expressions inside curly braces {} to be evaluated and inserted into the string dynamically.
2. **\n:** The \n adds a new line before the string output. This ensures the text starts on a new line.
3. **{self.name}:** Inside {}, the value of the self.name attribute (likely the student's name) is inserted dynamically. For example, if self.name = "Alice", this part of the string will be replaced with "Alice".
4. **'s Percentage:** A static part of the string, this will appear exactly as it is, providing context for the dynamic parts.
5. **{percentage:.2f}:** This embeds the value of the variable percentage into the string, formatting it to 2 decimal places.
    * The .2f part is a format specifier:
    * f stands for "fixed-point notation."
    * .2 means the number will be rounded to 2 decimal places.
    * **For example:**
    * If percentage = 85.666666, it will appear as 85.67.
    * If percentage = 92.0, it will appear as 92.00.
6. **%:**  The % symbol at the end is a static part of the string, indicating the percentage unit.


#### 2nd Version : Improved Program by implementing Error Handling
* In our 1st version there was no error handling. it means while entering student data if user makes mistakes and enters string or -ve number etc
  then this kinds of erros aren't handled by this 1st version of program.
* We have achieved it here in our 2nd Version

In [93]:
class Student:
    def percentage(self):
        # Taking student name input
        self.name = input("Enter the student's name: ")

        # Function to validate input
        def get_valid_marks(subject):
            while True:
                try:
                    marks = float(input(f"Enter score for {subject} (out of 100): "))
                    if 0 <= marks <= 100:
                        return marks  # Return valid marks
                    else:
                        print("Please enter marks between 0 and 100.")
                except ValueError:
                    print("Invalid input. Please enter a numeric value.")

        # Taking and validating marks for each subject
        self.maths = get_valid_marks("Maths")
        self.science = get_valid_marks("Science")
        self.physics = get_valid_marks("Physics")
        
        # Calculating percentage
        total_marks = self.maths + self.science + self.physics
        percentage = (total_marks / 300) * 100

        # Displaying the result
        print(f"\n{self.name}'s Percentage: {percentage:.2f}%")
# Creating an instance of the Student class
student1 = Student()
student2 = Student()
student3 = Student()

In [97]:
# Calling the percentage method
student1.percentage()

Enter the student's name:  David
Enter score for Maths (out of 100):  sjfnsghhhhhhhhhhhhhhhhhhhh


Invalid input. Please enter a numeric value.


Enter score for Maths (out of 100):  89
Enter score for Science (out of 100):  87
Enter score for Physics (out of 100):  67



David's Percentage: 81.00%


In [96]:
student2.percentage()

Enter the student's name:  Keval
Enter score for Maths (out of 100):  253677777777777777777777777


Please enter marks between 0 and 100.


Enter score for Maths (out of 100):  87
Enter score for Science (out of 100):  98
Enter score for Physics (out of 100):  95



Keval's Percentage: 93.33%


In [99]:
student3.percentage()

Enter the student's name:  Sid
Enter score for Maths (out of 100):  -567889543445556


Please enter marks between 0 and 100.


Enter score for Maths (out of 100):  92
Enter score for Science (out of 100):  84
Enter score for Physics (out of 100):  65



Sid's Percentage: 80.33%


#### How Above 2nd version of Program Works:
##### 1. Class Definition: 
* The **Student class** encapsulates the logic related to a student1,student2.etc.

##### 2. percentage Method:
* There is single method that takes user input for the student's name and marks in Maths, Science, and English,
  calculates percentage and print results.
    ##### 2.1 Validation Loop:
    * A helper function get_valid_marks is added to validate user input for each subject. It keeps prompting the user until a valid value (between 0 and 100) is entered.
    * If the input is invalid (not numeric or out of range), an appropriate message is displayed.
    ##### 2.2 Error Handling with try and except:
    * Added error handling for non-numeric inputs to avoid program crashes when the user enters invalid data like text.
    ##### 2.3 Reusable Validation Logic:
    * Instead of duplicating validation code for each subject, a single function handles all three subjects (Maths, Science, and Physics).


##### 3. Instance Creation: 
* An object of the Student class is created, and the percentage method is called ---> **student1.percentage()**

### 3rdVersion : Improving 2nd Version again and lets see how to make it more Optimized :
* Our **1st and 2nd version** program calculates the percentage for 3 students individually by creating five separate instances of the Student class. 
* While this works, it can be optimized for better space and time efficiency.
* In our 3rd version instead of 3 students we'll give facility to user to enter input total number of students for whom they want to calculate percentage.
* Also we will create just single instance s1, execute whole program just once and calculate percentage for as many students as user wants.

In [100]:
class Student_Optimized:
    def __init__(self,name,math,science,physics):
        self.name = name
        self.math = math
        self.science = science
        self.physics = physics
        
    def calculate_percentage(self):
        total_marks = self.math + self.science + self.physics
        percentage = (total_marks/300)*100
        return percentage
    
# Function to handle multiple students data   
def main():
    num_students = int(input("Enter the number of students: "))
    student_list = []
    
    #Taking user input
    for _ in range(num_students):
        name = input("\nEnter the student's name: ")
        def get_valid_marks(subject):
            while True:
                try:
                    marks = float(input(f"Enter score for {subject} (out of 100): "))
                    if 0 <= marks <= 100:
                        return marks  # Return valid marks
                    else:
                        print("Please enter marks between 0 and 100.")
                except ValueError:
                    print("Invalid input. Please enter a numeric value.")
                    
        math = get_valid_marks("Maths")
        science = get_valid_marks("Science")
        physics = get_valid_marks("Physics")
        
        # creating a s1 object
        s1 = Student_Optimized(name,math,science,physics)
        student_list.append(s1) # storing the s1 object input data to student_list
    
    # Displaying the result
    print("\nStudent's percentages:")
    for s1 in student_list:
        percentage = s1.calculate_percentage()
        print(f"{s1.name}'s percentage : {percentage:.2f}%")
        
# callin the main() function
main()

Enter the number of students:  2

Enter the student's name:  Keval
Enter score for Maths (out of 100):  95
Enter score for Science (out of 100):  97
Enter score for Physics (out of 100):  95

Enter the student's name:  Sid
Enter score for Maths (out of 100):  adhajfshkfjhskjgs


Invalid input. Please enter a numeric value.


Enter score for Maths (out of 100):  -654443333333


Please enter marks between 0 and 100.


Enter score for Maths (out of 100):  88
Enter score for Science (out of 100):  76
Enter score for Physics (out of 100):  87



Student's percentages:
Keval's percentage : 95.67%
Sid's percentage : 83.67%


#### Improvements Made :
##### 1. Single Student_Optimized Class Initialization (__ init __):

* Instead of using multiple calls to input inside the class, inputs are handled externally, and attributes are passed directly when creating the **s1** object. This improves modularity and separation of concerns.

##### 2. Handling Multiple Students Dynamically:

* Instead of creating multiple hardcoded instances (s1, s2, s3, etc.) in our previous program, here we store all s1 objects in a list **(student_list)** and iterate through it.

##### 3. Time Complexity:

* The input and calculations are streamlined using a loop. For n students, the time complexity is O(n) because we only loop through the input and calculations once.

##### 4. Space Complexity:

* By using a list to store student objects, we avoid multiple hardcoded variable names, making the program scalable for any number of students. Space complexity is O(n), where n is the number of students.

##### 5. Reusability:

* The Student_Optimized class has a dedicated calculate_percentage() method, which can be reused elsewhere, improving maintainability and scalability.


## 8. Understanding __ init __ method in Python 
#### 8.1 What is the **__ init __** Method in Python?
* The **__ init __** method in Python is a special method, also known as a constructor, used to initialize an object’s attributes when a class is instantiated. It is automatically called when you create a new instance of a class.

#### 8.2 Purpose of **__ init __** :
* ##### The **__ init __** method is primarily used to:
    1. Initialize Instance Attributes: Set up the initial state of an object by assigning values to its attributes.
    2. Run Setup Code: Perform any necessary setup or initialization tasks for the instance.
    
#### 8.3 Key Points about the **__ init __** Method
* #### Automatic Invocation:
    * **__ init __** is automatically invoked when an object is created from a class.
    * You don’t need to call it explicitly.
    
* #### Optional Parameters:
    * You can define parameters for the **__ init __** method to accept data when creating an object.
    * These parameters allow flexibility in initializing objects with different values.
    
* #### Instance-Specific Initialization:
    * Each object can have its own unique state because the attributes are initialized specifically for that object.

* #### Not Mandatory:
    * A class does not require an **__ init __** method if no initialization is needed.
* #### Default Return Value:
    * **__ init __** does not return anything; it always implicitly returns None.

* #### Self Reference:
    * self refers to the specific instance being initialized, allowing you to set instance attributes.



### 9. Understanding Design Philosophy of Python : "Everything in Python is an Object"
* ####  An integral part of Python's Object-Oriented Programming

In [73]:
class MyStudent:
    def percentage(self):
        # Displaying the type of self
        print(f"The data type of 'self' is: {type(self)}")

        # Taking user inputs
        self.name = input("Enter the student's name: ")
        self.maths = float(input("Enter marks for Maths (out of 100): "))
        self.science = float(input("Enter marks for Science (out of 100): "))
        self.english = float(input("Enter marks for English (out of 100): "))
        
        # Displaying the data type of self.name
        print(f"The data type of 'self.name' is: {type(self.name)}")
        print(f"The data type of 'self.maths' is: {type(self.maths)}")
        
        # Calculating percentage
        total_marks = self.maths + self.science + self.english
        percentage = (total_marks / 300) * 100
        
        # Printing the data type of percentage
        print(f"The data type of 'percentage' is: {type(percentage)}")
        print(f"The data type of 'total_marks' is: {type(total_marks)}")

        # Displaying the result
        print(f"\n{self.name}'s Percentage: {percentage:.2f}%")

# Creating an instance of the Student class
studentinput = MyStudent()

# Calling the percentage method
studentinput.percentage()

# Check the data type of class,an instance of class, the percentage function and 'percentage' when accessed from an instance

print(f"The data type of the 'def percentage' when accessed from the class is : {type(MyStudent.percentage)}")  # Accessing as a class attribute
print(f"The data type of 'studentinput.percentage()' method when accessed from instance is : {type(studentinput.percentage)}")  # Accessing as an instance attribute
print(f"The data type of an instace(studentinput) of a class MyStudent is : {type(studentinput)}")
print(f"The data type of 'class Mystudent' is : {type(MyStudent)}")

The data type of 'self' is: <class '__main__.MyStudent'>


Enter the student's name:  Siddhant Patel
Enter marks for Maths (out of 100):  92
Enter marks for Science (out of 100):  70
Enter marks for English (out of 100):  89


The data type of 'self.name' is: <class 'str'>
The data type of 'self.maths' is: <class 'float'>
The data type of 'percentage' is: <class 'float'>
The data type of 'total_marks' is: <class 'float'>

Siddhant Patel's Percentage: 83.67%
The data type of the 'def percentage' when accessed from the class is : <class 'function'>
The data type of 'studentinput.percentage()' method when accessed from instance is : <class 'method'>
The data type of an instace(studentinput) of a class MyStudent is : <class '__main__.MyStudent'>
The data type of 'class Mystudent' is : <class 'type'>


### Everything is an Object In Python:
* ##### All data and code are objects.
* ##### Objects are instances of classes, and their type can be determined using the type() function.
* ##### So Every object in Python belongs to a class that defines its behavior and properties.check the type of an object using type()
#### 1.The type of "self" :
* The data type of **self** is the instance of the class itself.
* In the context of the MyStudent class, when we use self, it refers to the current instance of the Student class. 
* **Therefore, the data type of self is <class '__main__.MyStudent'>** if the code is being run in the main module.
* **self** is a reference to the instance of the class, and **its data type will always be the class type in which it is used.**
* In this case, self is of type <class '__main__.MyStudent'>, indicating it is an instance of the MyStudent class defined in the __main__ module.

#### 2.Understanding the Type of the Function percentage :
* In Python, functions themselves have a data type, just like variables do. The data type of a function is <class 'function'>.
* The function def percentage in our MyStudent class is an instance method. Methods in Python are objects of type function. When we call the method using the class instance (e.g., studentinput.percentage()), it becomes a bound method, meaning it is bound to the specific instance (studentinput).
    * Print(type( **MyStudent.percentage** )) : This refers to the unbound method directly from the class and has the type <class 'function'>.
    * print(type( **studentinput.percentage** )): This refers to the bound method attached to the instance (studentinput) and has the type <class 'method'>.
    ##### Key Points:
    ##### 1.Unbound Method (<class 'function'>):
    * When you access the function directly from the class, it is treated as a plain function.
    ##### 2. Bound Method (<class 'method'>):
    * When the function is accessed via an instance, Python automatically binds it to that instance, turning it into a bound method.

    ##### Summary:
* The data type of the percentage function is **<class 'function'> when accessed from the class.**
* The data type of percentage becomes **<class 'method'> when accessed from an instance.**
* This distinction is essential for Python's object-oriented behavior.

#### 3. Instances are Objects of class they are created from:
* An instance of a class is an object of the class it was created from, and its data type is <class 'YourClassName'>. 
* **Every present instance of a class is also an object**, in our case instance **s1 is an object and its data type is <class '__main__.MyStudent'>.** However, if we access a method (function within the class) through the instance, it is a bound method, and its data type is <class 'method'>.
* for example : **studentinput.percentage()** its instance accessed using percentage method so **its type is <class 'method'>**

#### 4. Classes Are Objects
* **Classes themselves are objects of type <class 'type'>.** 
* This means:
    * You can dynamically create classes using type().
    * A class is an instance of the type metaclass, which defines how classes behave.

### Summary :
##### Everything in Python is an object.

##### 1.Even primitives like integers, strings, and floats are objects.
* Every object has a data type.
* The data type of an object is determined by its class.
##### 2.Functions are objects.
* Functions are instances of <class 'function'>.
##### 3.Classes are also objects.
* Classes are instances of <class 'type'>.
##### 4.Every instance of a class is an object.
* Its data type is <class 'CreatedClassName'>. (In our case s1 instance belongs to class we created which is MyClass)
##### 5.Methods are objects too.
* When a method is accessed through an instance, it becomes a bound method, and its data type is <class 'method'>.